x/cron/keeper/keeper.go

```diff
@@ -0,0 +1,264 @@
+	// set KeyTable if it has not already been set
+	if !ps.HasKeyTable() {
+		ps = ps.WithKeyTable(types.ParamKeyTable())
+	}
+
+	return &Keeper{
+		cdc:           cdc,
+		storeKey:      storeKey,
+		memKey:        memKey,
+		paramstore:    ps,
+		accountKeeper: accountKeeper,
+	}
+}
+
+func (k *Keeper) Logger(ctx sdk.Context) log.Logger {
+	return ctx.Logger().With("module", fmt.Sprintf("x/%s", types.ModuleName))
+}
+
+// ExecuteReadySchedules gets all schedules that are due for execution (with limit that is equals to Params.Limit)
+// and executes messages in each one
+// NOTE that errors in contract calls rollback all already executed messages
+func (k *Keeper) ExecuteReadySchedules(ctx sdk.Context) {
+	telemetry.ModuleMeasureSince(types.ModuleName, time.Now(), LabelExecuteReadySchedules)
+	schedules := k.getSchedulesReadyForExecution(ctx)
+
+	for _, schedule := range schedules {
+		err := k.executeSchedule(ctx, schedule)
+		recordExecutedSchedule(err, schedule)
+	}
+}
+
+// AddSchedule adds new schedule to execution for every block `period`.
+// First schedule execution is supposed to be on `now + period` block.
+func (k *Keeper) AddSchedule(ctx sdk.Context, name string, period uint64, msgs []types.MsgExecuteContract) error {
+	if k.scheduleExists(ctx, name) {
+		return fmt.Errorf("schedule already exists with name=%v", name)
+	}
+
+	schedule := types.Schedule{
+		Name:              name,
+		Period:            period,
+		Msgs:              msgs,
+		LastExecuteHeight: uint64(ctx.BlockHeight()), // let's execute newly added schedule on `now + period` block
+	}
+	k.storeSchedule(ctx, schedule)
+	k.changeTotalCount(ctx, 1)
+
+	return nil
+}
+
+// RemoveSchedule removes schedule with a given `name`
+func (k *Keeper) RemoveSchedule(ctx sdk.Context, name string) {
+	if !k.scheduleExists(ctx, name) {
+		return
+	}
+
+	k.changeTotalCount(ctx, -1)
+	k.removeSchedule(ctx, name)
+}
+
+// GetSchedule returns schedule with a given `name`
+func (k *Keeper) GetSchedule(ctx sdk.Context, name string) (*types.Schedule, bool) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.ScheduleKey)
+	bzSchedule := store.Get(types.GetScheduleKey(name))
+	if bzSchedule == nil {
+		return nil, false
+	}
+
+	var schedule types.Schedule
+	k.cdc.MustUnmarshal(bzSchedule, &schedule)
+	return &schedule, true
+}
+
+// GetAllSchedules returns all schedules
+func (k *Keeper) GetAllSchedules(ctx sdk.Context) []types.Schedule {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.ScheduleKey)
+
+	res := make([]types.Schedule, 0)
+
+	iterator := sdk.KVStorePrefixIterator(store, []byte{})
+	defer iterator.Close()
+
+	for ; iterator.Valid(); iterator.Next() {
+		var schedule types.Schedule
+		k.cdc.MustUnmarshal(iterator.Value(), &schedule)
+		res = append(res, schedule)
+	}
+
+	return res
+}
+
+func (k *Keeper) GetScheduleCount(ctx sdk.Context) int32 {
+	return k.getScheduleCount(ctx)
+}
+
+func (k *Keeper) StoreSchedule(ctx sdk.Context, schedule types.Schedule) {
```

Comment:
not sure what's happening here... you have `AddSchedule()`, `storeSchedule()` and `StoreSchedule()`, and the last one is only used in genesis. looks a bit weird?

Reply:
I used storeSchedule to store schedules with specified `LastExecuteHeight`, otherwise they get stored with current block height which can be not easy to manage, especially in tests

Reply:
Ok, so I removed the StoreSchedule, and now it's much clearer!